File: packages/cds-android/src/main/java/com/coinbase/cds/components/button/ButtonStyle.kt
package com.coinbase.cds.components.button

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import com.coinbase.cds.theme.CdsTheme

/**
 * Resolved container/content colors for a [ButtonVariant] — a flat lookup by variant, nothing more.
 *
 * The transparent variants use a true [Color.Transparent] container rather than the base `bg`
 * token. Painting `bg` only reads as "transparent" when the button happens to sit directly on the
 * screen's base background; this way it looks right on any surface a caller places it on, such as
 * a `bgSecondary` card.
 */
@Immutable
internal data class ButtonColors(val container: Color, val content: Color)

@Composable
internal fun buttonColors(variant: ButtonVariant, transparent: Boolean): ButtonColors =
    resolveButtonColors(variant, transparent, CdsTheme.colors)

internal fun resolveButtonColors(
    variant: ButtonVariant,
    transparent: Boolean,
    colors: com.coinbase.cds.theme.CdsColors,
): ButtonColors {
    return if (transparent) {
        when (variant) {
            ButtonVariant.Primary -> ButtonColors(Color.Transparent, colors.fgPrimary)
            ButtonVariant.Secondary -> ButtonColors(Color.Transparent, colors.fg)
            ButtonVariant.Tertiary -> ButtonColors(Color.Transparent, colors.fg)
            ButtonVariant.Positive -> ButtonColors(Color.Transparent, colors.fgPositive)
            ButtonVariant.Negative -> ButtonColors(Color.Transparent, colors.fgNegative)
            ButtonVariant.Inverse -> ButtonColors(Color.Transparent, colors.fg)
        }
    } else {
        when (variant) {
            ButtonVariant.Primary -> ButtonColors(colors.bgPrimary, colors.fgInverse)
            ButtonVariant.Secondary -> ButtonColors(colors.bgSecondary, colors.fg)
            ButtonVariant.Tertiary -> ButtonColors(colors.bgTertiary, colors.fg)
            ButtonVariant.Positive -> ButtonColors(colors.bgPositive, colors.fgInverse)
            ButtonVariant.Negative -> ButtonColors(colors.bgNegative, colors.fgInverse)
            ButtonVariant.Inverse -> ButtonColors(colors.bgInverse, colors.fgInverse)
        }
    }
}

/** Resolved size-derived metrics for a [ButtonSize]. */
@Immutable
internal data class ButtonMetrics(
    val paddingX: Dp,
    val paddingY: Dp,
    val radius: Dp,
    val iconSize: Dp,
    val font: TextStyle,
)

@Composable
internal fun buttonMetrics(size: ButtonSize): ButtonMetrics =
    resolveButtonMetrics(
        size = size,
        space = CdsTheme.space,
        borderRadius = CdsTheme.borderRadius,
        iconSize = CdsTheme.iconSize,
        typography = CdsTheme.typography,
    )

internal fun resolveButtonMetrics(
    size: ButtonSize,
    space: com.coinbase.cds.theme.CdsSpace,
    borderRadius: com.coinbase.cds.theme.CdsBorderRadius,
    iconSize: com.coinbase.cds.theme.CdsIconSize,
    typography: com.coinbase.cds.theme.CdsTypography,
): ButtonMetrics = when (size) {
    ButtonSize.Xs -> ButtonMetrics(space.x2, space.x0_75, borderRadius.radius700, iconSize.s, typography.label1)
    ButtonSize.S -> ButtonMetrics(space.x2, space.x1, borderRadius.radius700, iconSize.s, typography.headline)
    ButtonSize.M -> ButtonMetrics(space.x3, space.x1_5, borderRadius.radius900, iconSize.m, typography.headline)
    ButtonSize.L -> ButtonMetrics(space.x4, space.x2, borderRadius.radius900, iconSize.m, typography.headline)
}

File: packages/cds-ios/Sources/Components/ButtonStyle.swift
import SwiftUI

/// Visual/semantic variant — the five in the current Figma Button spec.
enum ButtonVariant { case primary, secondary, tertiary, positive, negative }

/// Size tier — the four sizes (`xs`/`s`/`m`/`l`) in the Figma Button spec.
enum ButtonSize { case xs, s, m, l }

/// Resolved container/content colors for a ``ButtonVariant``. Transparent variants use a true
/// `.clear` container so the button reads correctly on any surface.
struct ButtonColors {
    let container: Color
    let content: Color
}

func buttonColors(_ variant: ButtonVariant, transparent: Bool, theme: CDSTheme) -> ButtonColors {
    let colors = theme.colors
    if transparent {
        switch variant {
        case .primary: return ButtonColors(container: .clear, content: colors.fgPrimary)
        case .secondary: return ButtonColors(container: .clear, content: colors.fg)
        case .tertiary: return ButtonColors(container: .clear, content: colors.fg)
        case .positive: return ButtonColors(container: .clear, content: colors.fgPositive)
        case .negative: return ButtonColors(container: .clear, content: colors.fgNegative)
        }
    }
    switch variant {
    case .primary: return ButtonColors(container: colors.bgPrimary, content: colors.fgInverse)
    case .secondary: return ButtonColors(container: colors.bgSecondary, content: colors.fg)
    case .tertiary: return ButtonColors(container: colors.bgTertiary, content: colors.fg)
    case .positive: return ButtonColors(container: colors.bgPositive, content: colors.fgInverse)
    case .negative: return ButtonColors(container: colors.bgNegative, content: colors.fgInverse)
    }
}

/// Resolved size-derived metrics for a ``ButtonSize``. There is no `height` field: height falls
/// out of `paddingY` plus the font's line height.
struct ButtonMetrics {
    let paddingX: CGFloat
    let paddingY: CGFloat
    let radius: CGFloat
    let iconSize: CGFloat
    let font: CDSTextStyle
}

func buttonMetrics(_ size: ButtonSize, theme: CDSTheme) -> ButtonMetrics {
    let space = theme.spacing
    let radius = theme.radius
    let icon = theme.iconSize
    switch size {
    case .xs: return ButtonMetrics(paddingX: space.x2, paddingY: space.x0_75, radius: radius.r700, iconSize: icon.s, font: .label1)
    case .s: return ButtonMetrics(paddingX: space.x2, paddingY: space.x1, radius: radius.r700, iconSize: icon.s, font: .headline)
    case .m: return ButtonMetrics(paddingX: space.x3, paddingY: space.x1_5, radius: radius.r900, iconSize: icon.m, font: .headline)
    case .l: return ButtonMetrics(paddingX: space.x4, paddingY: space.x2, radius: radius.r900, iconSize: icon.m, font: .headline)
    }
}
